Closes #19523: Add counter‑cached counts & filters for DeviceType, ModuleType, and RackType

pheus opened the pull request:
<!--
    Thank you for your interest in contributing to NetBox! Please note that
    our contribution policy requires that a feature request or bug report be
    approved and assigned prior to opening a pull request. This helps avoid
    waste time and effort on a proposed change that we might not be able to
    accept.

    IF YOUR PULL REQUEST DOES NOT REFERENCE AN ISSUE WHICH HAS BEEN ASSIGNED
    TO YOU, IT WILL BE CLOSED AUTOMATICALLY.

    Please specify your assigned issue number on the line below.
-->
### Fixes: #19523

**Summary**  
Extend usage-count filtering via counter caches to cover **DeviceType**, **ModuleType**, and **RackType**.

**Implementation**
- Add counter‑cached integer fields:
  - `device_count` on **DeviceType**
  - `module_count` on **ModuleType**
  - `rack_count` on **RackType**
- Expose numeric lookups in **REST** and **GraphQL** (e.g., `device_count__gte=10`, `module_count=0`, `{rack_count: {gt: 2}`).
- Include migration to add fields and backfill counts.
- Add a test.

**Database Changes**  
Adds integer fields on the three models listed above.

jeremystretch commented on netbox/dcim/migrations/0218_devicetype_device_count.py:
```diff
@@ -0,0 +1,25 @@
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
```
We need to populate the field values as part of the migration as well (using a function called via `RunPython`), to ensure the counts are accurate. Without this, the counts all default to zero for existing objects.

pheus replied:
Thanks! Will do.

jeremystretch reviewed (changes requested):
Nice work @pheus!

pheus commented:
Thanks for the thoughtful review and the migration tip - much appreciated!

I’ve added a **data migration** to prefill `device_count` and `module_count`, and included a test to ensure we don’t double‑register the signal receivers. Happy to adjust anything further.

jeremystretch commented:
@pheus I think it makes sense to fold that in here as well, since it's a continuation of the same pattern. Thanks!

pheus commented on netbox/dcim/models/racks.py:
```diff
@@ -19,9 +19,11 @@
 from .device_components import PowerPort
-from .devices import Device, Module
+from .devices import Device
+from .modules import Module
```
_Note_: Unrelated import fix. Happy to drop this if you'd prefer to keep the PR focused.

pheus commented on netbox/dcim/models/racks.py:
```diff
@@ -509,7 +515,7 @@ def get_rack_units(self, user=None, face=DeviceFaceChoices.FACE_FRONT, exclude=N
         return [u for u in elevation.values()]
 
-    def get_available_units(self, u_height=1, rack_face=None, exclude=None, ignore_excluded_devices=False):
+    def get_available_units(self, u_height=1.0, rack_face=None, exclude=None, ignore_excluded_devices=False):
```
_Note_: Unrelated cleanup. Happy to drop this if you'd prefer to keep the PR focused.

pheus committed:
feat(dcim): Add device, module and rack count filters

Introduces `device_count`, `module_count` and `rack_count` filters to
enable queries based on the existence and count of the associated
device, module or rack instances.
Updates forms, filtersets, and GraphQL schema to support these filters,
along with tests for validation.

Fixes #19523

jeremystretch reviewed (approved):
Thanks @pheus, great work!
